aptos-move/block-executor/src/modules_and_scripts.rs

```diff
@@ -0,0 +1,407 @@
+        &self,
+        address: &AccountAddress,
+        module_name: &IdentStr,
+    ) -> VMResult<Option<usize>> {
+        Ok(self
+            .read_module_storage(address, module_name)?
+            .into_versioned()
+            .map(|(_, entry)| entry.bytes().len()))
+    }
+
+    fn fetch_module_metadata(
+        &self,
+        address: &AccountAddress,
+        module_name: &IdentStr,
+    ) -> VMResult<Vec<Metadata>> {
+        Ok(self
+            .get_existing_module_storage_entry_with_version(address, module_name)?
+            .1
+            .metadata()
+            .to_vec())
+    }
+
+    fn fetch_deserialized_module(
+        &self,
+        address: &AccountAddress,
+        module_name: &IdentStr,
+    ) -> VMResult<Arc<CompiledModule>> {
+        Ok(self
+            .get_existing_module_storage_entry_with_version(address, module_name)?
+            .1
+            .as_compiled_module())
+    }
+
+    fn fetch_verified_module(
+        &self,
+        address: &AccountAddress,
+        module_name: &IdentStr,
+    ) -> VMResult<Arc<Module>> {
+        let mut visited = HashSet::new();
+        self.traversed_published_dependencies(address, module_name, &mut visited)
+    }
+}
+
+impl<'a, T: Transaction, S: TStateView<Key = T::Key>, X: Executable> LatestView<'a, T, S, X> {
+    /// Returns the module storage entry built from the base (storage) view. If there is a
+    /// storage error, returns an error.
+    fn get_base_module_storage_entry(
+        &self,
+        key: &T::Key,
+    ) -> VMResult<Option<Arc<ModuleStorageEntry>>> {
+        self.get_raw_base_value(key)
+            .map_err(|e| e.finish(Location::Undefined))?
+            .map(|s| {
+                ModuleStorageEntry::from_state_value(self.runtime_environment, s).map(Arc::new)
+            })
+            .transpose()
+    }
+
+    /// Returns the module storage entry built from the current view. If it is not in
+    /// multi-version or non-sync data structures, fetches it from the base view.
+    fn read_module_storage_by_key(
+        &self,
+        key: &T::Key,
+    ) -> VMResult<ModuleStorageRead<ModuleStorageEntry>> {
+        match &self.latest_view {
+            ViewState::Sync(state) => {
+                // If the module read has been previously cached, return early.
+                if let Some(read) = state
+                    .captured_reads
+                    .borrow()
+                    .get_captured_module_storage_read(key)
+                {
+                    return Ok(read.clone());
+                }
+
+                // Otherwise, we need to go to the multi-version data structure to get it, and
+                // record under captured reads.
+                let read = state
+                    .versioned_map
+                    .code_storage()
+                    .module_storage()
+                    .get_or_else(key, self.txn_idx, || {
+                        self.get_base_module_storage_entry(key)
```

Comment:
this may destroy performance when every txn is trying to read 0x1::framework and all go into the base version and do deserialization/verification. just a note

Reply:
Yep, we should do it under lock, I think @gelash suggested that as well. We can just add a flag there, but what do we do with all other threads? Wait? Ideally we just let them do some other useful work while they wait.

Reply:
lock's implementation should put the thread to sleep and the resources can be used elsewhere in the big system, so it shouldn't be too bad.